Eigen::RowMajor cv2eigen assertion failed (+ naive fix)

- OpenCV => 4.2.0-dev
- Operating System / Platform => Debian 10 x86_64 GNU/Linux (4.19.0-6-amd64)
- Compiler => g++ 8.3.0

##### Detailed description

cv::cv2eigen doesn't work as I expected if the Eigen::Matrix is of type:
`Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>`.

I came up with a naive fix, but as I am relatively new to *both* OpenCV and Eigen (and especially at interfacing them), I thought I'd first open an issue and request a sanity check before going for a pull request. It seems entirely possible that I missed something obvious.

##### Steps to reproduce

```.cpp
cv::Mat_<double> image = cv::imread("../Lena.png", cv::IMREAD_GRAYSCALE);

Eigen::Matrix<double,Eigen::Dynamic, Eigen::Dynamic> A;
cv::cv2eigen(image, A); // Works as expected.

Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> B;
cv::cv2eigen(image, B); // Expected to work, but doesn't with the current opencv-master.
```

More precisely, it will compile, and then give at run time:

	terminate called after throwing an instance of 'cv::Exception'
	  what():  OpenCV(4.2.0-dev)  /usr/local/include/opencv4/opencv2/core/mat.inl.hpp:548: error: (-215:Assertion failed) total() == 0 || data != NULL in function 'Mat'

	Aborted

##### Naive fix:

I made a copy of an existing function in eigen.hpp (line 140 of that file), simply adding `Eigen::RowMajor` so that the argument now reads `Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst`. The body of the function is the same.
```cpp

// What I've added in eigen.hpp; seems to solve the problem
template<typename _Tp>  static inline
void cv2eigen( const Mat& src,
               Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst )
{
    dst.resize(src.rows, src.cols);
    if( !(dst.Flags & Eigen::RowMajorBit) )
    {
        const Mat _dst(src.cols, src.rows, traits::Type<_Tp>::value,
             dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
        if( src.type() == _dst.type() )
            transpose(src, _dst);
        else if( src.cols == src.rows )
        {
            src.convertTo(_dst, _dst.type());
            transpose(_dst, _dst);
        }
        else
            Mat(src.t()).convertTo(_dst, _dst.type());
    }
    else
    {
        const Mat _dst(src.rows, src.cols, traits::Type<_Tp>::value,
                 dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
        src.convertTo(_dst, _dst.type());
    }
}
```
Once that fct is defined inside the cv namespace, the programme compiles and runs, seemingly without issues.
```cpp
cv::cv2eigen(image, B); // Works with B.
cv::cv2eigen(image, A); // Works with A.

// Everything's fine
H5Easy::File output_hdf("out.hdf", H5Easy::File::Overwrite);
H5Easy::dump(output_hdf, "/A", A);
H5Easy::dump(output_hdf, "/B", B);

```

Cheers.

--- a/modules/core/include/opencv2/core/eigen.hpp
+++ b/modules/core/include/opencv2/core/eigen.hpp
@@ -287,6 +287,16 @@ void cv2eigen( const Mat& src,
     }
 }
 
+template<typename _Tp>  static inline
+void cv2eigen( const Mat& src,
+               Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst )
+{
+    dst.resize(src.rows, src.cols);
+    const Mat _dst(src.rows, src.cols, traits::Type<_Tp>::value,
+             dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
+    src.convertTo(_dst, _dst.type());
+}
+
 // Matx case
 template<typename _Tp, int _rows, int _cols> static inline
 void cv2eigen( const Matx<_Tp, _rows, _cols>& src,
@@ -305,6 +315,16 @@ void cv2eigen( const Matx<_Tp, _rows, _cols>& src,
                  dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
         Mat(src).copyTo(_dst);
     }
+}
+
+template<typename _Tp, int _rows, int _cols> static inline
+void cv2eigen( const Matx<_Tp, _rows, _cols>& src,
+               Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst )
+{
+    dst.resize(_rows, _cols);
+    const Mat _dst(_rows, _cols, traits::Type<_Tp>::value,
+                   dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
+    Mat(src).copyTo(_dst);
 }
 
 template<typename _Tp> static inline

--- a/modules/core/include/opencv2/core/eigen.hpp
+++ b/modules/core/include/opencv2/core/eigen.hpp
@@ -291,6 +291,7 @@ template<typename _Tp>  static inline
 void cv2eigen( const Mat& src,
                Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst )
 {
+    CV_CheckEQ(src.dims, 2, "");
     dst.resize(src.rows, src.cols);
     const Mat _dst(src.rows, src.cols, traits::Type<_Tp>::value,
              dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));
@@ -321,6 +322,7 @@ template<typename _Tp, int _rows, int _cols> static inline
 void cv2eigen( const Matx<_Tp, _rows, _cols>& src,
                Eigen::Matrix<_Tp, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>& dst )
 {
+    CV_CheckEQ(src.dims, 2, "");
     dst.resize(_rows, _cols);
     const Mat _dst(_rows, _cols, traits::Type<_Tp>::value,
                    dst.data(), (size_t)(dst.outerStride()*sizeof(_Tp)));

--- a/modules/core/test/test_mat.cpp
+++ b/modules/core/test/test_mat.cpp
@@ -2259,6 +2259,23 @@ TEST(Core_Eigen, eigen2cv_check_Mat_type)
     EXPECT_ANY_THROW(eigen2cv(eigen_A, d_mat));
     //EXPECT_EQ(CV_64FC1, d_mat.type());
 }
+
+TEST(Core_Eigen, cv2eigen_check_RowMajor)
+{
+    Mat A(2, 2, CV_32FC1, Scalar::all(0));
+    A.at<float>(0,0) = 1.0;
+    A.at<float>(0,1) = 2.0;
+    A.at<float>(1,0) = 3.0;
+    A.at<float>(1,1) = 4.0;
+
+    Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> eigen_A;
+    EXPECT_NO_THROW(cv2eigen(A, eigen_A));
+
+    ASSERT_EQ(eigen_A(0, 0), 1.0);
+    ASSERT_EQ(eigen_A(0, 1), 2.0);
+    ASSERT_EQ(eigen_A(1, 0), 3.0);
+    ASSERT_EQ(eigen_A(1, 1), 4.0);
+}
 #endif // HAVE_EIGEN
 
 #ifdef OPENCV_EIGEN_TENSOR_SUPPORT

--- a/modules/core/test/test_mat.cpp
+++ b/modules/core/test/test_mat.cpp
@@ -2273,12 +2273,12 @@ TEST(Core_Eigen, cv2eigen_check_RowMajor)
     Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> eigen_A;
     EXPECT_NO_THROW(cv2eigen(A, eigen_A));
 
-    ASSERT_EQ(eigen_A(0, 0), 1.0);
-    ASSERT_EQ(eigen_A(0, 1), 2.0);
-    ASSERT_EQ(eigen_A(1, 0), 3.0);
-    ASSERT_EQ(eigen_A(1, 1), 4.0);
-    ASSERT_EQ(eigen_A(2, 0), 5.0);
-    ASSERT_EQ(eigen_A(2, 1), 6.0);
+    ASSERT_EQ(1.0, eigen_A(0, 0));
+    ASSERT_EQ(2.0, eigen_A(0, 1));
+    ASSERT_EQ(3.0, eigen_A(1, 0));
+    ASSERT_EQ(4.0, eigen_A(1, 1));
+    ASSERT_EQ(5.0, eigen_A(2, 0));
+    ASSERT_EQ(6.0, eigen_A(2, 1));
 }
 #endif // HAVE_EIGEN
 

--- a/modules/core/test/test_mat.cpp
+++ b/modules/core/test/test_mat.cpp
@@ -2262,11 +2262,13 @@ TEST(Core_Eigen, eigen2cv_check_Mat_type)
 
 TEST(Core_Eigen, cv2eigen_check_RowMajor)
 {
-    Mat A(2, 2, CV_32FC1, Scalar::all(0));
+    Mat A(3, 2, CV_32FC1, Scalar::all(0));
     A.at<float>(0,0) = 1.0;
     A.at<float>(0,1) = 2.0;
     A.at<float>(1,0) = 3.0;
     A.at<float>(1,1) = 4.0;
+    A.at<float>(2,0) = 5.0;
+    A.at<float>(2,1) = 6.0;
 
     Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> eigen_A;
     EXPECT_NO_THROW(cv2eigen(A, eigen_A));
@@ -2275,6 +2277,8 @@ TEST(Core_Eigen, cv2eigen_check_RowMajor)
     ASSERT_EQ(eigen_A(0, 1), 2.0);
     ASSERT_EQ(eigen_A(1, 0), 3.0);
     ASSERT_EQ(eigen_A(1, 1), 4.0);
+    ASSERT_EQ(eigen_A(2, 0), 5.0);
+    ASSERT_EQ(eigen_A(2, 1), 6.0);
 }
 #endif // HAVE_EIGEN
 

